src/test/components/UrlManager.test.js

```diff
@@ -28,33 +35,83 @@ describe('UrlManager', function() {
           <UrlManager>Contents</UrlManager>
         </Provider>
       );
-    return { dispatch, getState, createUrlManager };
+
+    const waitUntilUrlSetupPhase = phase =>
+      waitUntilState(store, state => getUrlSetupPhase(state) === phase);
+
+    return { dispatch, getState, createUrlManager, waitUntilUrlSetupPhase };
   }
 
-  it('sets up the URL', function() {
-    const { getState, createUrlManager } = setup();
+  beforeEach(function() {
+    const profileJSON = createGeckoProfile();
+    const mockGetProfile = jest.fn().mockResolvedValue(profileJSON);
+
+    const geckoProfiler = {
+      getProfile: mockGetProfile,
+      getSymbolTable: jest
+        .fn()
+        .mockRejectedValue(new Error('No symbol tables available')),
+    };
+    window.fetch = jest
+      .fn()
+      .mockRejectedValue(new Error('No symbolication API in place'));
+    window.geckoProfilerPromise = Promise.resolve(geckoProfiler);
+  });
+
+  afterEach(function() {
+    delete window.geckoProfilerPromise;
```

Comment:
Is there any reason why we couldn't use `jest.spyOn(window, 'geckoProfilerPromise').mockImplementation(...);` instead of manually creating and deleting? I'm not sure if this is copied from another test but I think it would be nice to use the more automatic mechanism here.

Reply:
We can't use `jest.spyOn` here because `window` needs to have that property in jest’s `JSDOM` environment to be able to use that `spyOn` function. And since `geckoProfilerPromise` is a custom property, we can't  use it and have to implement it with `.fn()` instead.

The same goes for `fetch` as well. It's `window`'s property after all but it's still not implemented in `JSDOM`. That's why we can't use spyOn for this either :( See: https://github.com/jsdom/jsdom/issues/1724

Reply:
Ahhhh... now I remember this! Thanks for the explanation.